server/elasticsearch/src/main/java/org/hawkular/apm/server/elasticsearch/AnalyticsServiceElasticsearch.java

```diff
@@ -1091,223 +639,243 @@ protected void buildCommunicationSummaryStatistics(Map<String, CommunicationSumm
+        s.setCount(stat.getCount());
+        s.setFaultCount(stat.getCount() - missing.getDocCount());
+        return s;
+    }
 
+    private static CompletionTime toCompletionTime(SearchHit searchHit) {
         try {
-            for (int i = 0; i < completionTimes.size(); i++) {
-                CompletionTime ct = completionTimes.get(i);
-                String json = mapper.writeValueAsString(ct);
+            return mapper.readValue(searchHit.getSourceAsString(), CompletionTime.class);
+        } catch (IOException e) {
+            msgLog.errorFailedToParse(e);
+            return null;
+        }
+    }
 
-                if (msgLog.isTraceEnabled()) {
-                    msgLog.tracef("Storing fragment completion time: %s", json);
-                }
+    private static PrincipalInfo toPrincipalInfo(Terms.Bucket bucket) {
+        PrincipalInfo pi = new PrincipalInfo();
+        pi.setId(bucket.getKey());
+        pi.setCount(bucket.getDocCount());
+        return pi;
+    }
 
-                bulkRequestBuilder.add(client.getClient().prepareIndex(client.getIndex(tenantId),
-                        FRAGMENT_COMPLETION_TIME_TYPE, ct.getId()).setSource(json));
-            }
-        } catch (JsonProcessingException e) {
-            throw new StoreException(e);
+    private static SearchResponse getSearchResponse(SearchRequestBuilder request) {
+        SearchResponse response = request.execute().actionGet();
+        if (response.isTimedOut()) {
+            msgLog.warnQueryTimedOut();
         }
+        return response;
+    }
 
-        BulkResponse bulkItemResponses = bulkRequestBuilder.execute().actionGet();
+    private String getComponentTypeForBucket(Terms.Bucket typeBucket, Terms.Bucket parent) {
+        if (typeBucket.getKey().equalsIgnoreCase("consumer")) {
+            return "consumer";
+        } else if (typeBucket.getKey().equalsIgnoreCase("producer")) {
+            return "producer";
+        } else {
+            return parent.getKey();
+        }
+    }
 
-        if (bulkItemResponses.hasFailures()) {
+    private void doAddMetrics(CommunicationSummaryStatistics css, Stats duration, long docCount) {
+        css.setMinimumDuration((long)duration.getMin());
+        css.setAverageDuration((long)duration.getAvg());
+        css.setMaximumDuration((long)duration.getMax());
+        css.setCount(docCount);
+    }
 
-            if (msgLog.isTraceEnabled()) {
-                msgLog.trace("Failed to store fragment completion times to elasticsearch: "
-                        + bulkItemResponses.buildFailureMessage());
-            }
+    private long getTraceCompletionCount(String tenantId, Criteria criteria, boolean onlyFaulty) {
+        String index = client.getIndex(tenantId);
+        if (!refresh(index)) {
+            return 0;
+        }
 
-            throw new StoreException(bulkItemResponses.buildFailureMessage());
+        FilterBuilder filter = FilterBuilders.existsFilter("fault");
```

Comment:
This could be moved inside the "if (onlyFaulty) {" statement as not used elsewhere.